File: agent-driven-architecture-main/examples/kotlin/src/test/kotlin/adr/app/TotalityTest.kt
// ── test/app/totality — gate check C13, BOTH HALVES, by reflection ────────
// Every ToolResult case has a Verb entry and signs; every Verb entry has a case; and
// the NAME is the same in all three places (6.8). That mapping is what makes the one
// unchecked cast in Verb.signOf total, and what makes 17.6's "the gate keys off
// names" literally true.
//
// The handler half asks the same question one seam over: every declared `Effect` leaf
// has exactly ONE registered performer, no performer claims a leaf it does not own,
// and an orphan is diagnosed rather than dropped.
//
// `EffectSamples` BELOW IS THE GATE'S OWN TOTALITY LEDGER, and it is stated here
// rather than hidden: it is the ONE out-of-folder site a novel effect kind costs in
// this port. `declaredCases()` above is the same cost for a new VERB and pre-dates the
// split. Neither is a defect; both are equalities against a derivation off the live
// sealed hierarchy, which is what stops the checkers beneath them going vacuous.

package adr.app

import adr.Driver
import adr.blocks.analysis.AnalysisBlock
import adr.blocks.artifact.ArtifactBlock
import adr.blocks.escalation.EscalationBlock
import adr.blocks.triage.TriageBlock
import adr.contract.AnalysisEffect
import adr.contract.ArtifactEffect
import adr.contract.Command
import adr.contract.Effect
import adr.contract.EscalationEffect
import adr.contract.ToolResult
import adr.contract.TriageEffect
import adr.contract.TriageResult
import adr.spine.boundary.RecordingSink
import adr.spine.pure.EffectKey
import adr.spine.pure.EffectPerformer
import adr.spine.pure.Emit
import adr.spine.pure.KeyedEffect
import adr.spine.pure.ORPHAN_EFFECT
import adr.spine.pure.PerformMode
import adr.spine.pure.Performers
import adr.spine.pure.StepIndex
import adr.spine.pure.TicketId
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName
import kotlin.reflect.KClass
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** Walking a sealed hierarchy, on a constructed type — test sources are in scope too. */
private class Sealed {

    fun leaves(type: KClass<*>): List<KClass<*>> =
        if (type.sealedSubclasses.isEmpty()) listOf(type)
        else type.sealedSubclasses.flatMap { leaves(it) }

    fun verbName(type: KClass<*>): String =
        (type.simpleName ?: error("a sealed subclass is never anonymous: $type"))
            .replaceFirstChar { it.lowercase() }
}

/**
 * The C13 checker itself, on a constructed type. C13 is a question about VALUES, so
 * its fixture pair is a pair of INPUTS to one checker rather than a pair of files on
 * disk — same checker, two inputs, exactly like the TS port's `registryGaps`. The
 * ALLOW half runs it over the shipped registry; the BLOCK half pulls one verb out
 * and watches it deny. A check nobody has watched fail is not a check.
 */
private class Totality {

    fun gaps(cases: Set<String>, names: Set<String>): List<String> =
        (cases - names).map { "\"$it\" is a declared ToolResult case with no Verb entry" } +
            (names - cases).map { "\"$it\" is registered but declares no ToolResult case" }
}

class TotalityTest {

    private val registry = Wiring().wireApp(Env()).registry

    /** The spine's own two cases are not verbs — nobody calls them; the boundary mints them. */
    private val spineCases = setOf("unhandled", "refused")

    private fun declaredCases(): Set<String> =
        Sealed().leaves(ToolResult::class).map { Sealed().verbName(it) }.toSet() - spineCases

    @Test
    fun `C13 - every ToolResult case has a registry entry, and every entry has a case`() {
        val names = registry.keys.map { it.value }.toSet()

        assertEquals(emptyList(), Totality().gaps(declaredCases(), names))
        assertEquals(12, names.size, "six blocks, twelve verbs")
    }

    @Test
    fun `C13 BLOCK-TEST - a registry with a verb pulled out is DENIED`() {
        val thinned = registry.keys.map { it.value }.toSet() - "confirmSeal"

        assertEquals(
            listOf("\"confirmSeal\" is a declared ToolResult case with no Verb entry"),
            Totality().gaps(declaredCases(), thinned),
        )
    }

    @Test
    fun `C13 - the Command hierarchy mirrors the ToolResult hierarchy, name for name`() {
        val results = Sealed().leaves(ToolResult::class).map { Sealed().verbName(it) }.toSet()
        val commands = Sealed().leaves(Command::class).map { Sealed().verbName(it) }.toSet()
        assertEquals(results, commands)
    }

    @Test
    fun `C13 - every committed Command carries a name the registry knows`() {
        val authority = RunAuthority()
        val app = Wiring().wireApp(Env(world = World(), authority = authority))
        Driver().driveCanonicalSession(app, authority)
        Driver().human(app, ToolName("noSuchTool"))

        val committed = app.bus.records().flatMap { it.commands }
        assertTrue(committed.isNotEmpty())
        committed.forEach { command ->
            val known = command.tool in registry.keys ||
                command is Command.Unhandled ||
                command is Command.Refused
            assertTrue(known, "unsignable command: $command")
        }
    }

    @Test
    fun `ONE MECHANIC - a presentation verb and a domain verb have the SAME registration shape`() {
        val domain = registry.getValue(ToolName("setPriority"))
        val presentation = registry.getValue(ToolName("setPanel"))
        assertEquals(domain::class, presentation::class, "one tool mechanic, not two")
    }
}

/**
 * ONE SAMPLE PER DECLARED EFFECT LEAF, so the handler check can drive a REAL effect
 * through a REAL sink rather than assert about a table.
 *
 * The map is hand-written and that is the point: its key set is asserted EQUAL to the
 * set Kotlin reflection derives from the live sealed hierarchy, so a leaf added
 * anywhere in the system fails here — loudly — instead of leaving the checker below
 * matching a set that quietly stopped being the system's. That is the C7 rot, refused
 * in advance, and it is the same move `declaredCases()` above makes for ToolResult.
 *
 * It is ALSO the out-of-folder cost of a novel effect kind in this port. Counted, in
 * the composition root's receipt, rather than claimed away.
 */
private class EffectSamples {

    fun samples(): Map<String, Effect> = mapOf(
        "Diag" to Effect.Diag(Timestamp(1), "a diagnostic line"),
        "LogDecision" to TriageEffect.LogDecision(
            Timestamp(2),
            TicketId("4118"),
            TriageResult.Priority.High,
            null,
            null,
        ),
        "PageOncall" to EscalationEffect.PageOncall(Timestamp(3), TicketId("4118")),
        "DeliverArtifact" to ArtifactEffect.DeliverArtifact(Timestamp(4), emptyList()),
        "PublishConclusion" to AnalysisEffect.PublishConclusion(Timestamp(5), "a conclusion"),
    )

    /** Which leaves each registered performer OWNS. Asserted total against both the
     *  derived leaf set and the shipped assembly, so a new leaf or a new contributing
     *  block cannot be answered by an out-of-date map. */
    fun owners(): Map<String, Set<String>> = mapOf(
        "triage" to setOf("LogDecision"),
        "escalation" to setOf("PageOncall"),
        "artifact" to setOf("DeliverArtifact"),
        "analysis" to setOf("PublishConclusion"),
        "spine" to setOf("Diag"),
    )
}

/**
 * Gate check C13, HANDLER HALF — every declared effect kind has exactly one registered
 * performer, no performer claims more than its own, and an orphaned effect is
 * diagnosed rather than dropped.
 *
 * Same checker, two inputs, exactly like the registry half above: the ALLOW test runs
 * the SHIPPED assembly, the BLOCK test runs one built with a performer withheld. A
 * check nobody has watched fail is not a check — and the BLOCK half drives its effect
 * through the real [AppSink], because a test that called the floor directly would
 * prove the floor exists, not that anything reaches it.
 */
class EffectTotalityTest {

    private val world = World()
    private val env = Env(world = world)

    private fun leaves(): Set<String> =
        Sealed().leaves(Effect::class).mapNotNull { it.simpleName }.toSet()

    private fun withheld(log: MutableList<String>): Performers = Performers(
        listOf(
            TriageBlock().performer { log += it },
            ArtifactBlock().performer(env.delivery),
            AnalysisBlock().performer(env.relay),
            DiagPerformer(log).performer(),
        ),
    )

    @Test
    fun `C13 - the sample set is DERIVED from the live sealed hierarchy, and is not empty`() {
        val declared = leaves()
        assertTrue(
            declared.isNotEmpty(),
            "the Effect hierarchy walked to nothing — the derivation is vacuous",
        )
        assertEquals(declared, EffectSamples().samples().keys)
        // The known spellings, pinned: a rename that silently de-scoped the checker
        // would leave the two sets equal and both wrong.
        assertTrue(
            declared.containsAll(
                setOf("Diag", "LogDecision", "PageOncall", "DeliverArtifact", "PublishConclusion"),
            ),
            "a known effect kind vanished from the derivation: $declared",
        )
        // …and the owner map is total over the same derived set, so the
        // exactly-its-own-leaves test below cannot go vacuous either.
        assertEquals(declared, EffectSamples().owners().values.flatten().toSet())
    }

    @Test
    fun `C13 ALLOW - the shipped dispatcher performs every declared effect leaf`() {
        val log = mutableListOf<String>()
        val performers = Wiring().effectPerformers(env, log)
        val orphans = EffectSamples().samples()
            .filterValues { performers.perform(it) != null }
            .keys
        assertEquals(emptySet(), orphans, "an effect kind no registered performer claims")
        assertEquals(
            listOf("triage", "escalation", "artifact", "analysis", "spine"),
            performers.blocks(),
            "four blocks of six contribute, plus the spine's own Diag at the root",
        )
    }

    /**
     * THE OVER-CLAIM DENIAL. [Performers.perform] is first-claim-wins over an ORDERED
     * list, and every `narrow` is a lambda authored inside a block file that no lint
     * reads. A block whose narrow widened to the whole of `Effect` would swallow every
     * effect registered behind it — including the spine's own `Diag`, which is last —
     * and the ALLOW half above would stay green, because every leaf would still
     * "perform". So each performer is asked to REFUSE every sample outside its own
     * sub-union.
     *
     * Red-proven by widening escalation's narrow to fabricate an `EscalationEffect`
     * from anything: this test then names escalation, and nothing else changes.
     */
    @Test
    fun `C13 - each registered performer claims EXACTLY its own leaves`() {
        val log = mutableListOf<String>()
        val performers = Wiring().effectPerformers(env, log).claims()
        val owners = EffectSamples().owners()
        val samples = EffectSamples().samples()

        assertEquals(
            owners.keys,
            performers.map { it.block }.toSet(),
            "the owner map and the shipped assembly must name the same blocks",
        )

        val overclaimed = performers.flatMap { performer ->
            val mine = owners.getValue(performer.block)
            samples.filterKeys { it !in mine }
                .filterValues { performer.tryPerform(it) }
                .keys
                .map { "${performer.block} claimed `$it`, which is not its own" }
        }
        assertEquals(emptyList(), overclaimed, "a performer that over-claims hijacks the scan")
    }

    @Test
    fun `C13 BLOCK - a dispatcher with one performer withheld names the orphaned kind`() {
        val log = mutableListOf<String>()
        val orphan = withheld(log).perform(EffectSamples().samples().getValue("PageOncall"))

        assertEquals(
            "$ORPHAN_EFFECT `PageOncall`",
            orphan?.note,
            "the withheld kind must be named, not merely counted",
        )
        // …and every OTHER kind still performs, so the block half is denying one thing.
        assertNull(withheld(log).perform(EffectSamples().samples().getValue("LogDecision")))
    }

    @Test
    fun `C13 BLOCK - the orphan is DIAGNOSED at the real sink, never silent and never a crash`() {
        val log = mutableListOf<String>()
        val sink = RecordingSink(AppSink(withheld(log), DiagPerformer(log)))

        sink.perform(
            KeyedEffect(
                EffectKey(StepIndex(0), 0),
                EffectSamples().samples().getValue("PageOncall"),
            ),
            PerformMode.LIVE,
        )

        assertEquals(listOf("diag[3] $ORPHAN_EFFECT `PageOncall`"), log)
        assertEquals(emptyList(), world.pages, "the withheld port was NOT reached")
        assertEquals(1, sink.performed.size, "and the descriptor still crossed the seam")
    }

    @Test
    fun `C13 - REPLAY still touches nothing, dispatcher or not`() {
        val log = mutableListOf<String>()
        val sink = AppSink(Wiring().effectPerformers(env, log), DiagPerformer(log))

        sink.perform(
            KeyedEffect(
                EffectKey(StepIndex(0), 0),
                EffectSamples().samples().getValue("PageOncall"),
            ),
            PerformMode.REPLAY,
        )

        assertEquals(emptyList(), world.pages)
        assertEquals(emptyList(), log)
    }

    /**
     * TIER INDEPENDENCE — the reason the assembly sits BESIDE the verb registry.
     *
     * Handlers are assembled by `Wiring.effectPerformers`, not carried on
     * `BlockRegistration`. That fork is not pinned by any ratified decision, so it owes
     * a proof: a two-of-six DEEP_TIER app registers two blocks' VERBS and must still
     * perform every effect kind in the system. On `BlockRegistration` it would ship a
     * partial performer set and handler totality would be unstateable.
     *
     * The tier is asserted to be genuinely REDUCED in the same test, so a wiring that
     * silently ignored `env.verbs` could not satisfy it.
     */
    @Test
    fun `C13 - a DEEP_TIER wiring still performs every declared effect kind`() {
        val deepEnv = Env(world = World(), verbs = DEEP_TIER)
        val deep = Wiring().wireApp(deepEnv)
        val whole = Wiring().wireApp(Env(world = World()))
        assertTrue(
            deep.registry.size < whole.registry.size,
            "the tier must actually be reduced, or this proves nothing: " +
                "${deep.registry.size} vs ${whole.registry.size}",
        )

        val log = mutableListOf<String>()
        val performers = Wiring().effectPerformers(deepEnv, log)
        val orphans = EffectSamples().samples()
            .filterValues { performers.perform(it) != null }
            .keys
        assertEquals(emptySet(), orphans, "a reduced tier must still be TOTAL over effects")
    }

    @Test
    fun `THE ASSEMBLY REFUSES two blocks claiming one seam`() {
        val log = mutableListOf<String>()
        val emit = Emit<String> { line -> log += line }
        val twice: List<EffectPerformer<*>> = listOf(
            TriageBlock().performer(emit),
            TriageBlock().performer(emit),
        )
        val failure = assertFailsWith<IllegalStateException> { Performers(twice) }
        assertTrue(
            failure.message.orEmpty().contains("same name"),
            "the refusal must name what collided: ${failure.message}",
        )
    }
}

File: agent-driven-architecture-main/examples/typescript/test/app/totality.test.ts
// ── C13, BOTH HALVES — and the AI SDK binding that still works ─────────────
// Every "ok" ToolResult case has a Verb entry and a `sign` branch; every declared
// `Effect` kind has a registered handler; and the runtime binding drives the
// whole thing offline.
//
// THE TWO LEDGERS IN THIS FILE ARE THE GATE'S OWN, and they are the out-of-folder
// cost of a new case in each half: `EXPECTED` below for a new VERB (twelve
// entries, pre-dating the handler split) and `EXPECTED_EFFECTS` further down for
// a new EFFECT KIND. Both are mapped types over the live union's own
// discriminant, which is what stops the runtime checkers beneath them going
// vacuous — the exact failure C7's derivation shipped.

import { readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { MockLanguageModelV3 } from "ai/test";
import { describe, expect, it } from "vitest";
import { fold } from "../../src/app/assemble";
import type { Effect, OkResult } from "../../src/app/contract";
import { initialState } from "../../src/app/contract";
import type { Ports } from "../../src/app/wire";
import {
  ALL_BLOCKS,
  DEEP_TIER,
  effectHandlers,
  effectSink,
  offlinePorts,
  wireApp,
} from "../../src/app/wire";
import { analysis } from "../../src/blocks/analysis/register";
import { artifact } from "../../src/blocks/artifact/register";
import { consoleBlock } from "../../src/blocks/console/register";
import type { PageOncall } from "../../src/blocks/escalation/register";
import { escalation } from "../../src/blocks/escalation/register";
import { inbox } from "../../src/blocks/inbox/register";
import { triage } from "../../src/blocks/triage/register";
import { runTurn } from "../../src/spine/agent/loop";
import { signResult } from "../../src/spine/boundary/action";
import { handlerSink, movingClock } from "../../src/spine/boundary/in-memory";
import { authority, Signature } from "../../src/spine/pure/actor";
import type { Handlers } from "../../src/spine/pure/effect";
import { admit, ORPHAN_EFFECT } from "../../src/spine/pure/effect";
import type { Timestamp } from "../../src/spine/pure/ids";
import { keyedEffect } from "../../src/spine/pure/keyed-effect";
import type { ToolResultBase } from "../../src/spine/pure/tool-result";
import { refused, seal } from "../../src/spine/pure/tool-result";
import type { BlockOwnership } from "../gate/totality";
import { handlerGaps, ownershipGaps, UNREGISTERED_TOOL } from "../gate/totality";
import { AGENT_RUN, harness } from "../harness";
import { must } from "../support/must";

// A compile-time half: this table must name every "ok" tool in the system, and
// the compiler fails the build if the union grows past it (mapped type over the
// union's own discriminant).
const EXPECTED: Record<OkResult["tool"], true> = {
  setPriority: true,
  requestEscalation: true,
  confirmEscalation: true,
  focusTicket: true,
  setPanel: true,
  recordFinding: true,
  requestSeal: true,
  confirmSeal: true,
  recallAnalysis: true,
  publishAnalysis: true,
  noteDrop: true,
  noteFault: true,
};

const sig = new Signature("Agent", authority("agent-run-7f"));

describe("C13 — the verb-registry half", () => {
  it("every declared verb is registered, and every registered verb signs", () => {
    const { app } = harness();
    expect([...app.registry.keys()].sort()).toEqual(Object.keys(EXPECTED).sort());

    for (const verb of app.registry.values()) {
      expect(verb.kind === "Reversible" || verb.kind === "Irreversible").toBe(true);
      expect(typeof verb.describe).toBe("string");
      expect(verb.describe.length).toBeGreaterThan(0);
    }
  });

  it("the spine's own two cases sign too — a refusal is a decision (5.4)", () => {
    const { app } = harness();
    expect(signResult(app.registry, seal(refused("confirmSeal", "nope")), sig, "c9")).toEqual({
      outcome: "refused",
      tool: "confirmSeal",
      sig,
      id: "c9",
      reason: "nope",
    });
  });
});

const usage = {
  inputTokens: { total: 1, noCache: 1, cacheRead: undefined, cacheWrite: undefined },
  outputTokens: { total: 1, text: 1, reasoning: undefined },
};

describe("the runtime binding — the boundary is hooked onto onStepFinish", () => {
  it("runs a real generateText loop offline and folds what the AGENT asked for", async () => {
    const h = harness();
    let call = 0;
    const model = new MockLanguageModelV3({
      doGenerate: async () => {
        call += 1;
        if (call === 1) {
          return {
            content: [
              {
                type: "tool-call" as const,
                toolCallId: "t1",
                toolName: "setPriority",
                input: JSON.stringify({ ticket: "4118", level: "High" }),
              },
            ],
            finishReason: { unified: "tool-calls" as const, raw: undefined },
            usage,
            warnings: [],
          };
        }
        return {
          content: [{ type: "text" as const, text: "Priority set to High." }],
          finishReason: { unified: "stop" as const, raw: undefined },
          usage,
          warnings: [],
        };
      },
    });

    const out = await runTurn({
      model,
      prompt: "ticket 4118 looks urgent",
      boundary: h.app.boundary,
      registry: h.app.registry,
      dispatchers: h.app.dispatchers,
    });

    expect(out.steps).toBe(2);
    expect(h.app.boundary.state.triage.priority.get("4118")).toBe("High");
    // the loop forwarded an ACTION; the boundary produced the recorded result
    expect(must(h.app.bus.records()[0]).actions).toEqual([
      { tool: "setPriority", input: { ticket: "4118", level: "High" } },
    ]);
    expect(must(h.app.bus.records()[0]).commands[0]).toMatchObject({
      tool: "setPriority",
      sig: { by: "Agent" },
    });
    expect(h.sink.performed.map((k) => k.effect.kind)).toEqual(["LogDecision"]);
  });
});

// ── The floor under block dispatch (§6.5) ──────────────────────────────────
// `foldOk`'s ownership chain is type predicates, and TypeScript TRUSTS a
// predicate it cannot verify. A block whose `owns` goes stale therefore narrows
// the result to `never` at compile time while a real value flows through at
// runtime — tsc exit 0, eslint exit 0, every test green, and then the fold
// returned `undefined` and the caller died on `out.effects is not iterable`.
//
// A crash out of the one arm 6.5 says must never crash. This pins the fix: the
// unclaimed result folds like any unknown tool name — no transition, one
// diagnostic, one notice naming the tool. Total AND observable.
describe("an unclaimed result folds observably instead of crashing (§6.5)", () => {
  it("emits a diagnostic and a notice rather than returning undefined", () => {
    const state = initialState({ tickets: [{ id: "4118", body: "refund not received" }] });
    const sig = new Signature("Agent", AGENT_RUN);
    const stale = { outcome: "ok", tool: "resolveTicket", ticket: "4118" } as never;

    const out = fold(state, [stale], 1 as Timestamp, sig);

    expect(out.effects).toHaveLength(1);
    // The fold's output is ATTRIBUTED now (docs/DECISIONS.md:85): each effect
    // rides the committed result it came from, and the ONLY way to read what an
    // arm emitted is the admission rule itself — which passes a `Routine`
    // effect through untouched, licence or no licence.
    expect(admit(new Set(), out.effects)[0]).toMatchObject({ kind: "Diag" });
    const notices = out.state.spine.notices;
    expect(notices).toHaveLength(1);
    expect(notices[0]).toMatchObject({ kind: "Rejected", tool: "resolveTicket" });
    // no block slice advanced
    expect(out.state.triage).toEqual(state.triage);
  });
});

// ── THE HARDENING OVER THAT FLOOR — `owns` is derived, and watched ────────
// The floor above keeps an unclaimed result from crashing. It does not stop one
// EXISTING: `owns` was a hand-kept clause, and a verb added without it left the
// build green and the result unclaimed.
//
// TWO LAYERS CLOSE THAT, and neither is the other's spare. THE COMPILER holds
// the union half: each predicate is now derived from a claim table that is a
// mapped type over its own block's result union (`spine/pure/tool-result`), so a
// case declared and a case claimed are one edit — proven must-fail in both
// directions by the `owns-under-claim` and `owns-over-claim` fixtures in
// test/gate/exhaustiveness.test.ts. THIS CHECK holds the half no type can: that
// the union a block claims is still the set of verbs it REGISTERS. Between the
// two there is nowhere left for the stale half to live.
//
// The census is the root's own allowlist beside the six published predicates,
// and it is cross-checked against the block folders on disk — a seventh block
// joins by existing, rather than by being remembered here.
const BLOCKS_DIR = join(dirname(new URL(import.meta.url).pathname), "..", "..", "src", "blocks");

type Owns = (r: ToolResultBase) => boolean;

const OWNS: ReadonlyMap<string, Owns> = new Map<string, Owns>([
  [triage.name, triage.owns],
  [escalation.name, escalation.owns],
  [consoleBlock.name, consoleBlock.owns],
  [artifact.name, artifact.owns],
  [analysis.name, analysis.owns],
  [inbox.name, inbox.owns],
]);

const liveOwnership: readonly BlockOwnership[] = ALL_BLOCKS.map((registration) => ({
  block: registration.block,
  tools: registration.verbs.map((verb) => verb.name),
  owns: must(OWNS.get(registration.block)),
}));

/** One block's predicate replaced, everything else the shipped census. The
 *  block-halves below are INPUTS to the same checker the allow-half runs. */
const withOwns = (block: string, owns: Owns): readonly BlockOwnership[] =>
  liveOwnership.map((entry) => (entry.block === block ? { ...entry, owns } : entry));

describe("ownership totality — a block claims exactly the verbs it registers", () => {
  it("the census reads every block on disk, and the whole verb vocabulary", () => {
    // A census that had quietly stopped reading the tree would report silence,
    // which is the vacuous-check failure this repository has shipped before.
    const onDisk = readdirSync(BLOCKS_DIR).sort();
    expect(liveOwnership.map((b) => b.block).sort()).toEqual(onDisk);
    expect([...OWNS.keys()].sort()).toEqual(onDisk);
    // …and the vocabulary is this file's OWN compile-time ledger, derived rather
    // than pinned a second time, so a verb costs no new out-of-folder site here.
    expect(liveOwnership.flatMap((b) => b.tools).sort()).toEqual(Object.keys(EXPECTED).sort());
  });

  it("ALLOWS the shipped blocks — every `owns` matches its own registration", () => {
    expect(ownershipGaps(liveOwnership)).toEqual([]);
  });

  it("DENIES a block whose `owns` UNDER-CLAIMS its own verb table", () => {
    const stale = withOwns("console", (r) => r.outcome === "ok" && r.tool === "focusTicket");
    expect(ownershipGaps(stale)).toEqual([
      '"console" registers "setPanel" but its `owns` does not claim it',
    ]);
  });

  it("DENIES a block that claims a verb a SIBLING registers", () => {
    const greedy = withOwns(
      "inbox",
      (r) => inbox.owns(r) || (r.outcome === "ok" && r.tool === "setPriority"),
    );
    expect(ownershipGaps(greedy)).toEqual([
      '"inbox" claims "setPriority" but registers no verb of that name',
    ]);
  });

  it("DENIES a block that claims a name NOTHING in the system registers", () => {
    const phantom = withOwns(
      "triage",
      (r) => triage.owns(r) || (r.outcome === "ok" && r.tool === UNREGISTERED_TOOL),
    );
    expect(ownershipGaps(phantom)).toEqual([
      `"triage" claims "${UNREGISTERED_TOOL}" but registers no verb of that name`,
    ]);
  });

  it("DENIES a predicate blind to `outcome` — the spine's two cases are never a block's", () => {
    const blind = withOwns("triage", (r) => r.tool === "setPriority");
    expect(ownershipGaps(blind)).toEqual([
      '"triage" claims a result whose outcome is "unhandled" — the spine\'s own arm folds those, never a block',
      '"triage" claims a result whose outcome is "refused" — the spine\'s own arm folds those, never a block',
    ]);
  });
});

// ── C13, handler half — every effect kind has exactly one handler ─────────
// The COMPILE-TIME half is the table below: a mapped type over the live union's
// own discriminant, so an effect kind added anywhere in the system stops this
// file compiling until it is named here. That is what stops the runtime checker
// beneath it from going vacuous — the exact failure C7's derivation shipped, and
// it is also, stated plainly, the ONE out-of-folder site a novel effect kind
// costs in this port. The verb half above has always cost the same.
//
// The RUN-TIME half is `handlerGaps`, run twice over the SAME assembly: once
// whole, once with a handler pulled out. And the BLOCK half is driven through a
// real `Sink` built the way the app builds one, not by calling the floor
// function directly — a test that calls the floor proves the function exists,
// not that anything reaches it.
const EXPECTED_EFFECTS: Record<Effect["kind"], true> = {
  Diag: true,
  LogDecision: true,
  PageOncall: true,
  DeliverArtifact: true,
  PublishConclusion: true,
};

describe("C13 — handler totality", () => {
  const declared = Object.keys(EXPECTED_EFFECTS);
  const rig = (): { log: string[]; ports: Ports } => {
    const log: string[] = [];
    return { log, ports: offlinePorts((line) => log.push(line)) };
  };

  it("ALLOWS the shipped dispatcher — every declared kind has a handler", () => {
    expect(handlerGaps(declared, effectHandlers(rig().ports))).toEqual([]);
  });

  it("DENIES a dispatcher with one handler pulled out", () => {
    const thinned: Record<string, unknown> = { ...effectHandlers(rig().ports) };
    delete thinned.PageOncall;
    expect(handlerGaps(declared, thinned)).toEqual([
      '"PageOncall" is a declared Effect kind with no registered handler',
    ]);
  });

  it("DENIES a handler registered under a kind the union does not declare", () => {
    const strayed: Record<string, unknown> = {
      ...effectHandlers(rig().ports),
      SendCarrierPigeon: () => undefined,
    };
    expect(handlerGaps(declared, strayed)).toEqual([
      '"SendCarrierPigeon" has a registered handler but is not a declared Effect kind',
    ]);
  });

  it("the missing handler is DIAGNOSED at the real sink — never silent, never a crash", () => {
    const { log, ports } = rig();
    const thinned: Record<string, unknown> = { ...effectHandlers(ports) };
    delete thinned.PageOncall;
    // the real construction path: the same factory `effectSink` uses.
    const sink = handlerSink(thinned as unknown as Handlers<Effect>, (effect) =>
      ports.log(`[diag @${effect.at}] ${effect.note}`),
    );

    const orphan: PageOncall = {
      kind: "PageOncall",
      at: 7 as Timestamp,
      effectClass: "Irreversible",
      ticket: "4118",
    };
    sink.perform(keyedEffect(0, 0, orphan), "LIVE");

    expect(log).toEqual([`[diag @7] ${ORPHAN_EFFECT} \`PageOncall\``]);
    // the on-call port was NOT reached, and nothing threw
    expect(log.some((line) => line.includes("pager"))).toBe(false);
  });

  it("REPLAY still touches nothing, dispatcher or not (G9)", () => {
    const { log, ports } = rig();
    const sink = effectSink(ports);
    sink.perform(
      keyedEffect(0, 0, {
        kind: "PageOncall",
        at: 7 as Timestamp,
        effectClass: "Irreversible",
        ticket: "4118",
      } satisfies PageOncall),
      "REPLAY",
    );
    expect(log).toEqual([]);
  });

  // ── TIER INDEPENDENCE, the reason handlers sit BESIDE the verb registry ───
  // Handlers are assembled by `effectHandlers(ports)`, not carried on
  // `BlockRegistration`. That fork was chosen because handler totality must hold
  // whatever tier is wired: a two-of-six DEEP_TIER app registers two blocks'
  // VERBS and still performs every effect kind in the system. On
  // `BlockRegistration` the same app would ship a partial handler set and the
  // totality rule would be unstateable. This is the assertion that keeps the
  // reason true instead of merely argued — and it proves the tier really is
  // reduced, so a wiring that silently ignored `verbs` could not satisfy it.
  it("a DEEP_TIER wiring still ships a TOTAL handler set", () => {
    const { ports } = rig();
    const clock = movingClock(1000 as Timestamp, 7 as Timestamp);
    const deep = wireApp({ clock, sink: effectSink(ports), verbs: DEEP_TIER });
    const whole = wireApp({ clock, sink: effectSink(ports), verbs: ALL_BLOCKS });

    expect(deep.registry.size).toBeLessThan(whole.registry.size);
    expect(handlerGaps(declared, effectHandlers(ports))).toEqual([]);
  });
});
